src/frontend/Ast.ml

```diff
@@ -54,31 +54,35 @@ type ('e, 'f) expression =
   | TupleExpr of 'e list
 [@@deriving sexp, hash, compare, map, fold]
 
-type ('m, 'f) expr_with = {expr: (('m, 'f) expr_with, 'f) expression; emeta: 'm}
-[@@deriving sexp, compare, map, hash, fold]
+type ('m, 'f, 'p) expr_with =
+  {expr: (('m, 'f, 'p) expr_with, 'f, 'p) expression; emeta: 'm}
+[@@deriving sexp, compare, hash]
 
 (** Untyped expressions, which have location_spans as meta-data *)
 type located_meta = {loc: (Location_span.t[@sexp.opaque] [@compare.ignore])}
-[@@deriving sexp, compare, map, hash, fold]
+[@@deriving sexp, compare, hash]
 
-type untyped_expression = (located_meta, unit) expr_with
-[@@deriving sexp, compare, map, hash, fold]
+(** Uninhabited type. This is supplied as an argument to the Promotion node in
+    untyped ASTs so the compiler can deduce it is impossible *)
+type empty = | [@@deriving sexp, compare, hash]
```

Comment:
There's `Core.Nothing.t`, why create a new type?

Reply:
Ah, I searched around in stdlib and came up empty, neglected to look as closely through Core